File: distribution/python/dqops/client/models/check_container_model.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..models.schedule_enabled_status_model import ScheduleEnabledStatusModel
from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.check_search_filters import CheckSearchFilters
    from ..models.delete_stored_data_queue_job_parameters import (
        DeleteStoredDataQueueJobParameters,
    )
    from ..models.effective_schedule_model import EffectiveScheduleModel
    from ..models.quality_category_model import QualityCategoryModel


T = TypeVar("T", bound="CheckContainerModel")


@_attrs_define
class CheckContainerModel:
    """Model that returns the form definition and the form data to edit all data quality checks divided by categories.

    Attributes:
        categories (Union[Unset, List['QualityCategoryModel']]): List of all data quality categories that contain data
            quality checks inside.
        effective_schedule (Union[Unset, EffectiveScheduleModel]): Model of a configured schedule (enabled on connection
            or table) or schedule override (on check). Describes the CRON expression and the time of the upcoming execution,
            as well as the duration until this time.
        effective_schedule_enabled_status (Union[Unset, ScheduleEnabledStatusModel]):
        partition_by_column (Union[Unset, str]): The name of the column that partitioned checks will use for the time
            period partitioning. Important only for partitioned checks.
        run_checks_job_template (Union[Unset, CheckSearchFilters]): Target data quality checks filter, identifies which
            checks on which tables and columns should be executed.
        data_clean_job_template (Union[Unset, DeleteStoredDataQueueJobParameters]):
        can_edit (Union[Unset, bool]): Boolean flag that decides if the current user can edit the check.
        can_run_checks (Union[Unset, bool]): Boolean flag that decides if the current user can run checks.
        can_delete_data (Union[Unset, bool]): Boolean flag that decides if the current user can delete data (results).
    """

    categories: Union[Unset, List["QualityCategoryModel"]] = UNSET
    effective_schedule: Union[Unset, "EffectiveScheduleModel"] = UNSET
    effective_schedule_enabled_status: Union[Unset, ScheduleEnabledStatusModel] = UNSET
    partition_by_column: Union[Unset, str] = UNSET
    run_checks_job_template: Union[Unset, "CheckSearchFilters"] = UNSET
    data_clean_job_template: Union[Unset, "DeleteStoredDataQueueJobParameters"] = UNSET
    can_edit: Union[Unset, bool] = UNSET
    can_run_checks: Union[Unset, bool] = UNSET
    can_delete_data: Union[Unset, bool] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        categories: Union[Unset, List[Dict[str, Any]]] = UNSET
        if not isinstance(self.categories, Unset):
            categories = []
            for categories_item_data in self.categories:
                categories_item = categories_item_data.to_dict()

                categories.append(categories_item)

        effective_schedule: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.effective_schedule, Unset):
            effective_schedule = self.effective_schedule.to_dict()

        effective_schedule_enabled_status: Union[Unset, str] = UNSET
        if not isinstance(self.effective_schedule_enabled_status, Unset):
            effective_schedule_enabled_status = (
                self.effective_schedule_enabled_status.value
            )

        partition_by_column = self.partition_by_column
        run_checks_job_template: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.run_checks_job_template, Unset):
            run_checks_job_template = self.run_checks_job_template.to_dict()

        data_clean_job_template: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.data_clean_job_template, Unset):
            data_clean_job_template = self.data_clean_job_template.to_dict()

        can_edit = self.can_edit
        can_run_checks = self.can_run_checks
        can_delete_data = self.can_delete_data

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if categories is not UNSET:
            field_dict["categories"] = categories
        if effective_schedule is not UNSET:
            field_dict["effective_schedule"] = effective_schedule
        if effective_schedule_enabled_status is not UNSET:
            field_dict["effective_schedule_enabled_status"] = (
                effective_schedule_enabled_status
            )
        if partition_by_column is not UNSET:
            field_dict["partition_by_column"] = partition_by_column
        if run_checks_job_template is not UNSET:
            field_dict["run_checks_job_template"] = run_checks_job_template
        if data_clean_job_template is not UNSET:
            field_dict["data_clean_job_template"] = data_clean_job_template
        if can_edit is not UNSET:
            field_dict["can_edit"] = can_edit
        if can_run_checks is not UNSET:
            field_dict["can_run_checks"] = can_run_checks
        if can_delete_data is not UNSET:
            field_dict["can_delete_data"] = can_delete_data

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.check_search_filters import CheckSearchFilters
        from ..models.delete_stored_data_queue_job_parameters import (
            DeleteStoredDataQueueJobParameters,
        )
        from ..models.effective_schedule_model import EffectiveScheduleModel
        from ..models.quality_category_model import QualityCategoryModel

        d = src_dict.copy()
        categories = []
        _categories = d.pop("categories", UNSET)
        for categories_item_data in _categories or []:
            categories_item = QualityCategoryModel.from_dict(categories_item_data)

            categories.append(categories_item)

        _effective_schedule = d.pop("effective_schedule", UNSET)
        effective_schedule: Union[Unset, EffectiveScheduleModel]
        if isinstance(_effective_schedule, Unset):
            effective_schedule = UNSET
        else:
            effective_schedule = EffectiveScheduleModel.from_dict(_effective_schedule)

        _effective_schedule_enabled_status = d.pop(
            "effective_schedule_enabled_status", UNSET
        )
        effective_schedule_enabled_status: Union[Unset, ScheduleEnabledStatusModel]
        if isinstance(_effective_schedule_enabled_status, Unset):
            effective_schedule_enabled_status = UNSET
        else:
            effective_schedule_enabled_status = ScheduleEnabledStatusModel(
                _effective_schedule_enabled_status
            )

        partition_by_column = d.pop("partition_by_column", UNSET)

        _run_checks_job_template = d.pop("run_checks_job_template", UNSET)
        run_checks_job_template: Union[Unset, CheckSearchFilters]
        if isinstance(_run_checks_job_template, Unset):
            run_checks_job_template = UNSET
        else:
            run_checks_job_template = CheckSearchFilters.from_dict(
                _run_checks_job_template
            )

        _data_clean_job_template = d.pop("data_clean_job_template", UNSET)
        data_clean_job_template: Union[Unset, DeleteStoredDataQueueJobParameters]
        if isinstance(_data_clean_job_template, Unset):
            data_clean_job_template = UNSET
        else:
            data_clean_job_template = DeleteStoredDataQueueJobParameters.from_dict(
                _data_clean_job_template
            )

        can_edit = d.pop("can_edit", UNSET)

        can_run_checks = d.pop("can_run_checks", UNSET)

        can_delete_data = d.pop("can_delete_data", UNSET)

        check_container_model = cls(
            categories=categories,
            effective_schedule=effective_schedule,
            effective_schedule_enabled_status=effective_schedule_enabled_status,
            partition_by_column=partition_by_column,
            run_checks_job_template=run_checks_job_template,
            data_clean_job_template=data_clean_job_template,
            can_edit=can_edit,
            can_run_checks=can_run_checks,
            can_delete_data=can_delete_data,
        )

        check_container_model.additional_properties = d
        return check_container_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties

File: dqops/src/main/java/com/dqops/services/check/mapping/models/CheckContainerModel.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */
package com.dqops.services.check.mapping.models;

import com.dqops.checks.DefaultRuleSeverityLevel;
import com.dqops.checks.comparison.AbstractComparisonCheckCategorySpecMap;
import com.dqops.core.jobqueue.jobs.data.DeleteStoredDataQueueJobParameters;
import com.dqops.metadata.id.HierarchyIdModel;
import com.dqops.metadata.search.CheckSearchFilters;
import com.dqops.utils.docs.generators.SampleValueFactory;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import io.swagger.annotations.ApiModel;
import lombok.Data;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Model that returns the form definition and the form data to edit all data quality checks divided by categories.
 */
@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@ApiModel(value = "CheckContainerModel", description = "Model that returns the form definition and the form data to edit all data quality checks divided by categories.")
public class CheckContainerModel {
    @JsonPropertyDescription("List of all data quality categories that contain data quality checks inside.")
    private List<QualityCategoryModel> categories = new ArrayList<>();

    @JsonPropertyDescription("Model of configured schedule enabled on the check container.")
    private EffectiveScheduleModel effectiveSchedule;

    @JsonPropertyDescription("State of the effective scheduling on the check container.")
    private ScheduleEnabledStatusModel effectiveScheduleEnabledStatus;

    @JsonPropertyDescription("The name of the column that partitioned checks will use for the time period partitioning. Important only for partitioned checks.")
    private String partitionByColumn;

    @JsonPropertyDescription("Configured parameters for the \"check run\" job that should be pushed to the job queue in order to start the job.")
    private CheckSearchFilters runChecksJobTemplate;

    @JsonPropertyDescription("Configured parameters for the \"data clean\" job that after being supplied with a time range should be pushed to the job queue in order to remove stored results connected with this check container")
    private DeleteStoredDataQueueJobParameters dataCleanJobTemplate;

    /**
     * Boolean flag that decides if the current user can edit the check.
     */
    @JsonPropertyDescription("Boolean flag that decides if the current user can edit the check.")
    private boolean canEdit;

    /**
     * Boolean flag that decides if the current user can run checks.
     */
    @JsonPropertyDescription("Boolean flag that decides if the current user can run checks.")
    private boolean canRunChecks;

    /**
     * Boolean flag that decides if the current user can delete data (results).
     */
    @JsonPropertyDescription("Boolean flag that decides if the current user can delete data (results).")
    private boolean canDeleteData;

    public CheckContainerModel() {
    }

    /**
     * Iterates over categories and removes checks that are already configured. This method is used by the rule mining module.
     */
    public void dropConfiguredChecks() {
        for (QualityCategoryModel categoryModel : this.getCategories()) {
            categoryModel.dropConfiguredChecks();
        }
    }

    /**
     * Drops empty categories that do not have any checks.
     */
    public void dropEmptyCategories() {
        ArrayList<QualityCategoryModel> copyOfCategories = new ArrayList<>(this.categories);
        for (QualityCategoryModel categoryModel : copyOfCategories) {
            if (categoryModel.getChecks().isEmpty()) {
                this.categories.remove(categoryModel);
            }
        }
    }

    /**
     * Changes the default rule severity that is proposed in the check editor to a given severity level.
     * It is used by the profiling default checks to disable proposing rule configuration, and focus on data quality assessment.
     * @param defaultRuleSeverityLevel Default rule severity level to assign to all checks.
     */
    public void changeDefaultSeverityLevel(DefaultRuleSeverityLevel defaultRuleSeverityLevel) {
        for (QualityCategoryModel categoryModel : this.getCategories()) {
            for (CheckModel checkModel : categoryModel.getChecks()) {
                checkModel.setDefaultSeverity(defaultRuleSeverityLevel);
            }
        }
    }

    /**
     * Collects the hierarchy ids of all configured checks in the model.
     * @return Hierarchy ids of all configured checks.
     */
    public List<HierarchyIdModel> collectConfiguredChecksHierarchyIds() {
        List<HierarchyIdModel> resultList = new ArrayList<>();

        for (QualityCategoryModel categoryModel : this.getCategories()) {
            for (CheckModel checkModel : categoryModel.getChecks()) {
                if (checkModel.isConfigured() && checkModel.getCheckSpec() != null && checkModel.getCheckSpec().getHierarchyId() != null) {
                    HierarchyIdModel hierarchyIdModel = checkModel.getCheckSpec().getHierarchyId().toHierarchyIdModel();
                    resultList.add(hierarchyIdModel);
                }
            }
        }

        return resultList;
    }

    /**
     * Removes the "comparisons" category, because we cannot copy these checks without creating a new comparison configuration.
     */
    public void removeComparisonCategory() {
        this.categories.removeIf(categoryModel -> Objects.equals(categoryModel.getCategory(), AbstractComparisonCheckCategorySpecMap.COMPARISONS_CATEGORY_NAME));
    }

    public static class CheckContainerModelSampleFactory implements SampleValueFactory<CheckContainerModel> {
        @Override
        public CheckContainerModel createSample() {
            return new CheckContainerModel() {{
                setCategories(List.of(new QualityCategoryModel.QualityCategoryModelSampleFactory().createSample()));
            }};
        }
    }
}
